File: WADV.Core/MessageService.vb
' WinUP Adventure Game Engine Core Framework
' Game Message Service Class
' This is the message service class

Imports System.Threading
Imports System.Collections.Concurrent
Imports WADV.Core.ReceiverList

''' <summary>
''' 游戏消息循环
''' </summary>
Friend NotInheritable Class MessageService
    Private _status As Boolean
    Private Shared _self As MessageService
    Private ReadOnly _receiverThread As Thread
    Private ReadOnly _messageList As ConcurrentQueue(Of String)
    Friend Shared ReadOnly LastMessage(49) As Char

    ''' <summary>
    ''' 获取消息循环的状态
    ''' </summary>
    ''' <value></value>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Friend ReadOnly Property Status As Boolean
        Get
            Return _status
        End Get
    End Property

    ''' <summary>
    ''' 发送一条消息
    ''' </summary>
    ''' <param name="message">消息内容</param>
    Friend Sub SendMessage(message As String)
        Monitor.Enter(_messageList)
        _messageList.Enqueue(message)
        Monitor.Pulse(_messageList)
        Monitor.Exit(_messageList)
    End Sub

    Private Sub New()
        _messageList = New ConcurrentQueue(Of String)
        _receiverThread = New Thread(AddressOf MessageContent)
        _receiverThread.Name = "[系统]消息循环线程"
        _receiverThread.Priority = ThreadPriority.AboveNormal
        _receiverThread.IsBackground = True
    End Sub

    Private Sub MessageContent()
        Dim message As String = Nothing
        SyncLock (_messageList)
            While _status
                While Not _messageList.IsEmpty
                    If Not _messageList.TryDequeue(message) Then Throw New System.Exception("从消息队列中获取消息失败")
#If DEBUG Then
                    Debug.WriteLine(DateTime.Now.ToString("HH:mm:ss,fff ") & message)
#End If
                    MessageReceiverList.Broadcast(message)
                    SyncLock LastMessage
                        For i As Integer = 0 To message.Length - 1
                            LastMessage(i) = message(i)
                        Next
                        For i As Integer = message.Length To 49
                            LastMessage(i) = Nothing
                        Next
                        Monitor.PulseAll(LastMessage)
                    End SyncLock
                End While
                Monitor.Wait(_messageList)
            End While
        End SyncLock
    End Sub

    ''' <summary>
    ''' 获取消息循环的唯一实例
    ''' </summary>
    ''' <returns></returns>
    ''' <remarks></remarks>
    Friend Shared Function GetInstance() As MessageService
        If _self Is Nothing Then _self = New MessageService
        Return _self
    End Function

    ''' <summary>
    ''' 启动消息循环
    ''' </summary>
    ''' <remarks></remarks>
    Friend Sub Start()
        _status = True
        _receiverThread.Start()
    End Sub

    ''' <summary>
    ''' 终止消息循环
    ''' </summary>
    ''' <remarks></remarks>
    Friend Sub [Stop]()
        _status = False
    End Sub

End Class

File: WADV.Core/System/MessageService.cs
using System;
using System.Collections.Concurrent;
using System.Threading;
using System.Diagnostics;
using WADV.Core.Exception;

namespace WADV.Core.System {
    internal sealed class MessageService {
        private static readonly MessageService SingleInstance = new MessageService();
        private readonly Thread _messageLooper;
        private readonly ConcurrentQueue<Message> _messages;
        private readonly char[] _lastMessage;
        private bool _status;

        private MessageService() {
            _status = false;
            _messages = new ConcurrentQueue<Message>();
            _lastMessage = new char[64];
            _messageLooper = new Thread(MessageLooperContent) {
                Name = "[SYSTEM] Messenger",
                IsBackground = true
            };
        }

        /// <summary>
        /// Instance of Message service
        /// </summary>
        /// <returns></returns>
        internal static MessageService Instance() {
            return SingleInstance;
        }

        /// <summary>
        /// Get or set status of Message service
        /// </summary>
        internal bool Status {
            get { return _status; }
            set {
                if (value == _status) return;
                _status = value;
                if (_status) _messageLooper.Start();
            }
        }

        /// <summary>
        /// Send a new message
        /// </summary>
        /// <param name="message"></param>
        internal void Send(Message message) {
            Monitor.Enter(_messages);
            _messages.Enqueue(message);
            Monitor.Pulse(_messages);
            Monitor.Exit(_messages);
        }

        private void MessageLooperContent() {
            lock (_messages) {
                while (_status) {
                    while (!_messages.IsEmpty) {
                        if (!_messages.TryDequeue(out var message)) {
                            throw GameException.New(ExceptionType.MessageDequeueFailed)
                                .Value("Message", GameException.NoValue)
                                .At("MessageService")
                                .How("Deque message failed")
                                .Save();
                        }
#if DEBUG
                        Debug.WriteLine(
                            DateTime.Now.ToString(
                                $"HH:mm:ss,fff[{message.Mask}] {message.Content}: {message.Parameter.ToString()} & {message.NumericalParameter}"));
#endif
                        Configuration.Receivers.MessageReceivers.Map(message);
                        lock (_lastMessage) {
                            int i;
                            for (i = 0; i < message.Content.Length; i++) {
                                _lastMessage[i] = message.Content[i];
                            }
                            _lastMessage[i] = '\0';
                            Monitor.PulseAll(_lastMessage);
                        }
                    }
                    Monitor.Wait(_messages);
                }
            }
        }
    }
}